File: core/session/audit.py
import json
import os
from collections.abc import Mapping
from datetime import datetime, timezone
from pathlib import Path

OUTPUT_PREVIEW_LIMIT = 1024


class ToolAuditLogger:
    """Append-only JSONL audit log of every tool call."""

    def __init__(self, project_path: str) -> None:
        self._path = Path(project_path) / ".orchestrator" / "tool_audit.jsonl"

    @property
    def path(self) -> Path:
        return self._path

    def record(
        self,
        *,
        persona: str,
        tool: str,
        args: Mapping[str, object],
        output: str,
        duration_ms: int,
        exit_status: int = 0,
        error: str | None = None,
    ) -> None:
        preview = output[:OUTPUT_PREVIEW_LIMIT]
        row: dict[str, object] = {
            "ts": _utc_iso_now(),
            "persona": persona,
            "tool": tool,
            "args": dict(args),
            "exit_status": exit_status,
            "duration_ms": duration_ms,
            "output_truncated": len(output) > OUTPUT_PREVIEW_LIMIT,
            "output_preview": preview,
        }
        if error is not None:
            row["error"] = error

        self._path.parent.mkdir(parents=True, exist_ok=True)
        line = json.dumps(row, ensure_ascii=False, default=str) + "\n"
        with open(self._path, "a", encoding="utf-8") as fp:
            fp.write(line)
            fp.flush()
            os.fsync(fp.fileno())


def _utc_iso_now() -> str:
    """UTC ISO-8601 timestamp with millisecond precision, e.g. 2026-05-22T19:30:00.123Z."""
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S") + f".{now.microsecond // 1000:03d}Z"

File: src/core/session/audit.ts
// Tool-call audit log (V1/06) — the only safety net for autonomous, no-confirmation tool calls
// (CLAUDE.md). Appends ONE JSON line per call to projects/<active>/.orchestrator/tool_audit.jsonl
// from the single dispatch choke point (V1/02) — tools never write their own rows. Ports
// core/session/audit.py, renaming the legacy `persona` field to `phase`.
//
// Append-only + flushed (fsync) per row: a kill mid-tool leaves a partial LAST line at worst, never
// a torn earlier row. Only a ~1 KB output PREVIEW is stored — full stdout/build logs stay out of
// the log (full-output persistence is deferred). The fsync-per-line durability lives in the shared
// appendJsonlLine writer, which the sibling events log (V5/04) uses too.

import path from 'node:path';

import { appendJsonlLine } from './append-jsonl-line.js';
import type { ToolCallRecord } from './tool-call-record.type.js';

/** Only the first ~1 KB of a tool's result is kept on the row (never the full output). */
export const OUTPUT_PREVIEW_LIMIT = 1024;

/**
 * Append one audit row for a dispatched tool call. Builds the row shape, then hands it to the shared
 * appendJsonlLine writer, which creates `<projectPath>/.orchestrator/` on first write and fsyncs the
 * line so prior rows are always intact on disk.
 */
export function appendAuditRow(projectPath: string, record: ToolCallRecord): void {
  const file = path.join(projectPath, '.orchestrator', 'tool_audit.jsonl');

  const output = record.output;
  const row: Record<string, unknown> = {
    ts: record.ts, // UTC ISO-8601 ms
    phase: record.phase, // NOT `persona` — the legacy Python name is not carried over
    tool: record.tool,
    args: record.args,
    exit_status: record.exitStatus, // 0 / real code / -1 for any failure
    duration_ms: record.durationMs,
    output_truncated: output.length > OUTPUT_PREVIEW_LIMIT,
    output_preview: output.slice(0, OUTPUT_PREVIEW_LIMIT),
    error: record.error, // null on success
  };
  if (record.metadata !== undefined) {
    row['metadata'] = record.metadata;
  }
  // A sub-agent's own tool call (V5/01) carries its id so lineage traces back to the sub-agent; a
  // master-phase call omits the field entirely (never a null placeholder).
  if (record.subagentId !== undefined) {
    row['subagent_id'] = record.subagentId;
  }

  appendJsonlLine(file, row);
}
